### PR TITLE
Add Comparison, Equivalence, Predicate

### PR DESCRIPTION
**Pull request by garyb**
I'll add some comments before merging, but just wanted some feedback first.

Do we think these all belong together? `Predicate`, `Equivalence` etc. could actually live in their own packages potentially.

### CODE REVIEW

**Comment by paf31** [comment]
:+1: I like it.

**Comment by joneshf** [comment]
Very nice!

File: src/Data/Equivalence.purs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+instance contravariantEquivalence :: Contravariant Equivalence where
+  (>$<) f g = Equivalence (\a b -> runEquivalence g (f a) (f b))
+
+instance semigroupEquivalence :: Semigroup (Equivalence a) where
+  (<>) (Equivalence p) (Equivalence q) = Equivalence (\a b -> p a b && q a b)
+
+instance monoidEquivalence :: Monoid (Equivalence a) where
+  mempty = Equivalence (\_ _ -> true)
+
+defaultEquivalence :: forall a. (Ord a) => Equivalence a
```
Comment (joneshf) [inline src/Data/Equivalence.purs]: You only need `Eq a` here right?

Comment (garyb) [inline src/Data/Equivalence.purs]: Oops, yep. Thanks.

**Comment by garyb** [comment]
Added some docs now, and moved the module to `Data.Functor.Contravariant` - I'll be making a repo for invariant functors soon too, so figured it'd make sense to group them both under `Data.Functor`.

Thoughts? And how do the docs look? @paf31 @joneshf

**Comment by paf31** [comment]
Looks great. Does it make sense to move `Compose` out into its own library? It has other instances which could be useful, like `Functor` and `Applicative`.

**Comment by garyb** [comment]
Sure, I can do that.

I felt similarly about `Comparison` and `Equivalence`, but figure I'll leave those for now, and perhaps when we dismantle the `Prelude` they can live with `Ord` and `Eq` or something.

**Comment by joneshf** [comment]
I thought `Compose` was already a thing! :) I agree with moving it to its own library. Or maybe just put it in `purescript-control`? 

I don't really understand the namespace change, but then I never have understood the namespaces.

File: src/Data/Functor/Contravariant/Compose.purs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+newtype Compose f g a = Compose (f (g a))
+
+runCompose :: forall f g a. Compose f g a -> f (g a)
+runCompose (Compose f) = f
+
+instance functorCompose :: (Contravariant f, Contravariant g) => Functor (Compose f g) where
+  (<$>) f (Compose x) = Compose ((f >$<) >$< x)
+
+-- | Composition of a covariant and contravariant functor.
+newtype ComposeFC f g a = ComposeFC (f (g a))
```
Comment (joneshf) [inline src/Data/Functor/Contravariant/Compose.purs]: Is there a reason for both `ComposeFC` and `ComposeCF`? Overlapping instances or some such?

Comment (garyb) [inline src/Data/Functor/Contravariant/Compose.purs]: Good point, having both versions might be unnecessary. Will give it a try after moving this out.

**Comment by paf31** [comment]
Sorry, I didn't think about the possible overlapping instances issue. I guess things are almost never going to be both covariant and contravariant, but it is a possible issue. Even if they are, the two instances should be identical, right? Maybe we just have to stick the two instances in different modules.

**Pull request merged**